### PR TITLE
feat: implement describe subcommand

### LINKED ISSUE #62: Implement `templates describe <url>`
### TL;DR

We want users to be able to find out information about a template, like:

 - The description
 - The set of inputs: which are required, what are their descriptions, what are their names

One possible design: we add a new subcommand `templates describe` that downloads a template, parses it's spec file, and prints out the metadata about the template in a human-friendly format.

Another idea: formalize the convention that "README.md must exist and contain usage instructions for the template, expected to be a few hundred words long at least" 

CC @verbanicm who mentioned this.

### Detailed design

_No response_

### Alternatives considered

_No response_

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by sailorlqh**
fix #62

### CODE REVIEW

File: templates/commands/describe/describe.go
Comment (sethvargo) [inline templates/commands/describe/describe.go:102]: I don't think you want this line

File: templates/commands/describe/describe.go
Comment (sethvargo) [inline templates/commands/describe/describe.go:107]:
```suggestion

	_, templateDir, err := templatesource.Download(ctx, rp.fs, rp.describeTempDirBase, c.flags.Source, c.flags.GitProtocol)
```

File: templates/commands/describe/describe.go
Comment (sethvargo) [inline templates/commands/describe/describe.go:153]: This logic seems like it could be extracted into a helper for more independent testing.

File: templates/commands/describe/describe.go
Comment (drevell) [inline templates/commands/describe/describe.go:132]: I'm not sure this line is necessary, it doesn't really convey any information

**Review by drevell — CHANGES_REQUESTED** [review]

File: templates/commands/describe/describe.go
Comment (sethvargo) [reply templates/commands/describe/describe.go:102]: The function signature already has this as a named return parameter, so this is creating a local shadow of the same variable name.

Comment (drevell) [reply templates/commands/describe/describe.go:102]: Ah, right.

File: templates/commands/describe/describe.go
Comment (sailorlqh) [reply templates/commands/describe/describe.go:132]: I was thinking this can work as something like a Header

File: templates/commands/describe/describe.go
Comment (sailorlqh) [reply templates/commands/describe/describe.go:152]: Yeah, I thought the same, I was planing on doing that in the next PR. I want to implement the feature first so this PR don't get blocked by the refactoring PR while you are out. 
I can work on the refactor PR and rebase this PR on that.

**Comment by sailorlqh** [comment]
Hold until another PR for refactoring is merged.

File: templates/commands/describe/describe.go
Comment (drevell) [reply templates/commands/describe/describe.go:132]: Yeah, I see. In this case I think it might end up being confusing to the user. I was a bit confused when I saw it, like "is there supposed to be something after the colon? I already know this is a template."

File: templates/commands/describe/describe.go
Comment (sailorlqh) [reply templates/commands/describe/describe.go:52]: I changed the following line to match the updated text after your fix, it this ok?

File: templates/commands/describe/describe_test.go
Comment (sailorlqh) [reply templates/commands/describe/describe_test.go:118]: I think this way is better, it can use the existing `common.WriteAllDefaultMode` to write the file to tempDir, so we don't need to write another function.

File: templates/commands/describe/describe.go
Comment (drevell) [reply templates/commands/describe/describe.go:52]: LGTM

File: templates/commands/describe/describe_test.go
Comment (drevell) [reply templates/commands/describe/describe_test.go:118]: I just realized that your way also lets us write a test case for spec.yaml being nonexistent. Could you add that (if you agree)?

File: templates/common/specutil/spec.go
Comment (drevell) [inline templates/common/specutil/spec.go:42]: I'm not sure about the name of this function. Particularly the word "Parse". Usually parsing means going from text to an object. This is going in the other direction (object -> text).

Also, it doesn't really describe the whole spec (it doesn't include the "steps"). So maybe a name like SpecFieldsForDescribe()?

Also, it doesn't seem to be used except for in the `describe` subcommand, so maybe it doesn't belong in common?

File: templates/common/specutil/spec.go
Comment (drevell) [inline templates/common/specutil/spec.go:43]: This might be little more understandable if it was a [][]string instead of a []string. Then each line of text could be it's own list, rather than using the assumption that every index multiple of 2 starts a new line.

File: templates/common/specutil/spec.go
Comment (drevell) [inline templates/common/specutil/spec.go:81]: I think these empty lines should have `//` otherwise this won't be seen as one big comment and it won't all be included in the automatically generated godocs.

File: templates/commands/describe/describe.go
Comment (drevell) [inline templates/commands/describe/describe.go:127]: Does this really need to be c.attrList? Could it be a local variable instead (just `attrList`)? Generally we should try to avoid storing state where it's not needed, which increases complexity.

**Review by drevell — CHANGES_REQUESTED** [review]

File: templates/commands/describe/describe.go
Comment (sailorlqh) [reply templates/commands/describe/describe.go:127]: the purpose of this is for doing diff in the tests. It doesn't need to be c.attrList, but this makes tests easier.

File: templates/commands/describe/describe_test.go
Comment (sailorlqh) [reply templates/commands/describe/describe_test.go:120]: Test case added

File: templates/common/specutil/spec.go
Comment (sailorlqh) [reply templates/common/specutil/spec.go]: Done.

**Review by drevell — APPROVED** [review]

**Pull request merged**
